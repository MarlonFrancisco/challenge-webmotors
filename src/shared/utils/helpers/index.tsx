import React from "react";
import { ThemeProvider } from "styled-components";
import { theme } from "../../styled";

type WithThemeProps = {};

export const WithTheme: React.FC<WithThemeProps> = ({ children }) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);
