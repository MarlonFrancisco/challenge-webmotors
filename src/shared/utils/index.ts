export type Options = {
  withoutPrefix: boolean;
  splitted: boolean;
};

export const formatCurrency = (
  value = 0,
  options?: Options
): string | string[] => {
  const formatted = new Intl.NumberFormat("pt-BR", {
    style: "currency",
    currency: "BRL",
  }).format(value);
  let apllyOptions: string | string[] = formatted;

  if (!options) {
    return formatted;
  }

  if (options.withoutPrefix) {
    apllyOptions = apllyOptions.replace("R$", "").trim();
  }

  if (options.splitted) {
    apllyOptions = apllyOptions.split(",");
  }

  return apllyOptions;
};
