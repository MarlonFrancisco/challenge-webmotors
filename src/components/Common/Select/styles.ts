import styled from "styled-components";
import { Theme } from "./../../../shared/@types";

export const Container = styled.div`
  display: flex;
  border: 1px solid
    ${(props: { theme: Theme }) => props.theme.palette.text.primary};
  padding: 6px;
  flex-basis: 100%;
  height: 30px;
  align-items: center;
  position: relative;
`;

export const SelectButton = styled.button`
  width: 100%;
  display: flex;
  justify-content: space-between;
  border: none;
  background: none;
`;

export const SelectBody = styled.div`
  display: flex;
  background: #fff;
  flex-direction: column;
  position: absolute;
  top: 30px;
  left: 0;
  width: 100%;
  border: 1px solid #b4a6a0;
  border-radius: 3px;
  display: ${({ open }: { open: boolean }) => (open ? "block" : "none")};
  z-index: 1;
`;

export const SelectItem = styled.button`
  border: none;
  background: #fff;
  padding: 10px 16px;
  font-size: 0.875rem;
  width: 100%;
  text-align: left;
  color: ${({ theme }: { theme: Theme }) => theme.palette.text.primary};

  &:hover {
    background: ${({ theme }: { theme: Theme }) => theme.palette.primary.main};
    color: #fff;
  }
`;
