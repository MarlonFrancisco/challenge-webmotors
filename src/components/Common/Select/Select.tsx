import React, { useState, useRef, useEffect, memo } from "react";
import { Typography } from "../Typography";
import { Container, SelectBody, SelectItem, SelectButton } from "./styles";
import { FaCaretDown, FaCaretUp } from "react-icons/fa";
import { useTheme } from "styled-components";

type Option = { Name: string; ID: number };

export type SelectProps = {
  LeftComponent?: React.ReactNode;
  items: Option[];
  onChange: (id: number) => void;
  value?: number;
  placeholder?: string;
};

const Select: React.FC<SelectProps> = ({
  LeftComponent,
  items,
  onChange,
  value,
  placeholder,
}) => {
  const theme = useTheme();
  const refContainer = useRef<HTMLDivElement>();
  const [open, setOpen] = useState(false);

  const handleChangeSelectStatus = (status: boolean) => setOpen(status);

  const handleSelectItem = (item: Option) => {
    handleChangeSelectStatus(false);
    onChange(item.ID);
  };

  useEffect(() => {
    const handleClickOutside = (e: MouseEvent) => {
      const element = e.target as HTMLElement;
      if (open && !refContainer.current?.contains(element)) {
        handleChangeSelectStatus(false);
      }
    };

    document.addEventListener("mousedown", handleClickOutside);

    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, [open]);
  return (
    <Container ref={refContainer as React.MutableRefObject<HTMLDivElement>}>
      <SelectButton onClick={() => handleChangeSelectStatus(!open)}>
        <div style={{ display: "flex" }}>
          {LeftComponent && LeftComponent}
          <Typography size={12} color="secondary" weight={600}>
            {items.find((item) => item.ID === value)?.Name ??
              placeholder ??
              "Todos"}
          </Typography>
        </div>

        {open ? (
          <FaCaretDown color={theme.palette.text.secondary} />
        ) : (
          <FaCaretUp color={theme.palette.text.secondary} />
        )}
      </SelectButton>

      <SelectBody className="select__body" open={open}>
        {!!items.length && (
          <SelectItem
            onClick={() => handleSelectItem({ Name: "Todos", ID: 0 })}
          >
            Todos
          </SelectItem>
        )}
        {items.map((item) => (
          <SelectItem key={item.ID} onClick={() => handleSelectItem(item)}>
            {item.Name}
          </SelectItem>
        ))}
        {!items.length && (
          <SelectItem onClick={() => handleChangeSelectStatus(false)}>
            Nenhum valor disponivel
          </SelectItem>
        )}
      </SelectBody>
    </Container>
  );
};

export default memo(Select);
