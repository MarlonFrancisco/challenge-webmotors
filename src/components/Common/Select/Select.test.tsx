import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Select, { SelectProps } from "./Select";
import { WithTheme } from "../../../shared/utils/helpers";

const Wrapper = (props: SelectProps) => (
  <WithTheme>
    <Select {...props} />
  </WithTheme>
);

describe("Select", () => {
  let props: SelectProps;

  beforeEach(() => {
    props = {
      items: [
        {
          Name: "Test",
          ID: 1,
        },
      ],
      onChange: jest.fn(),
    };
  });

  it("should render without an error", () => {
    const { container } = render(<Wrapper {...props} />);
    expect(container).toMatchSnapshot();
  });

  it("should render props", () => {
    const { queryByText } = render(
      <Wrapper
        {...props}
        LeftComponent={<p>LeftComponent</p>}
        placeholder="Cores"
      />
    );

    expect(queryByText("Cores")).not.toBeNull();
    expect(queryByText("LeftComponent")).not.toBeNull();
  });

  it("should render option select", () => {
    const { queryByText } = render(<Wrapper {...props} />);

    expect(queryByText("Test")).not.toBeNull();
  });

  it("should call onChange function", () => {
    const { getByText } = render(<Wrapper {...props} />);

    const triggeredEvent = fireEvent.click(getByText("Test"));

    expect(triggeredEvent).toBeTruthy();
    expect(props.onChange).toHaveBeenCalled();
  });
});
