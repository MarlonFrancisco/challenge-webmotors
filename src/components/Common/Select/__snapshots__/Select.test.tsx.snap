// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select should render without an error 1`] = `
<div>
  <div
    class="sc-bdnylx bzntss"
  >
    <button
      class="sc-gtssRu cSEATA"
    >
      <div
        style="display: flex;"
      >
        <h1
          style="font-size: 12px; color: rgb(132, 131, 131); font-weight: 600; text-align: center;"
        >
          Todos
        </h1>
      </div>
      <svg
        color="#848383"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        style="color: rgb(132, 131, 131);"
        viewBox="0 0 320 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M288.662 352H31.338c-17.818 0-26.741-21.543-14.142-34.142l128.662-128.662c7.81-7.81 20.474-7.81 28.284 0l128.662 128.662c12.6 12.599 3.676 34.142-14.142 34.142z"
        />
      </svg>
    </button>
    <div
      class="sc-dlnjPT irWVGS select__body"
    >
      <button
        class="sc-hKFyIo dQSCIi"
      >
        Todos
      </button>
      <button
        class="sc-hKFyIo dQSCIi"
      >
        Test
      </button>
    </div>
  </div>
</div>
`;
